core/src/main/scala/kafka/server/ClientRequestQuotaManager.scala

```diff
@@ -36,23 +36,36 @@ class ClientRequestQuotaManager(private val config: ClientQuotaManagerConfig,
     exemptSensor.record(value)
   }
 
-  def maybeRecordAndThrottle(request: RequestChannel.Request, sendResponseCallback: Int => Unit): Unit = {
+  /**
+    * Records that a user/clientId changed some metric being throttled (produced/consumed bytes, request processing
+    * time, etc.) If quota has been violated, return throttle time in milliseconds. Throttle time calculation may be
+    * overridden by sub-classes.
+    * @param request client request
+    * @return Number of milliseconds to throttle in case of quota violation. Zero otherwise
+    */
+  def maybeRecord(request: RequestChannel.Request): Int = {
     if (request.apiRemoteCompleteTimeNanos == -1) {
       // When this callback is triggered, the remote API call has completed
       request.apiRemoteCompleteTimeNanos = time.nanoseconds
     }
 
     if (quotasEnabled) {
-      val quotaSensors = getOrCreateQuotaSensors(request.session.sanitizedUser, request.header.clientId)
-      request.recordNetworkThreadTimeCallback = Some(timeNanos => recordNoThrottle(quotaSensors, nanosToPercentage(timeNanos)))
-
-      recordAndThrottleOnQuotaViolation(
-          quotaSensors,
-          nanosToPercentage(request.requestThreadTimeNanos),
-          sendResponseCallback)
-    } else {
-      sendResponseCallback(0)
+      request.recordNetworkThreadTimeCallback = Some(timeNanos => recordNoThrottle(getOrCreateQuotaSensors(request),
+        nanosToPercentage(timeNanos)))
     }
+    super.maybeRecord(request, nanosToPercentage(request.requestThreadTimeNanos))
+  }
+
+  /**
+    * A convenience method that calls maybeRecord() and maybeThrottle() if throttling is needed.
+    * @param request client request
+    * @return A tuple that consists of throttling delay in milliseconds and an optional ThrottledChannel object.
+    */
+  def maybeRecordAndThrottle(request: RequestChannel.Request): (Int, Option[ThrottledChannel]) = {
+    var throttleTimeMs = maybeRecord(request)
+    if (throttleTimeMs > 0) {
```

Comment:
It is probably not necessary to check `throttleTimeMs > 0` since `maybeThrottle()` will check it anyway.

Reply:
Done